### PR TITLE
[gen2] supply secure

### PR DESCRIPTION
**Pull request by keeramis**
### Description

Gen 2 supply secure

Platforms - Photon/P1/Electron

### Steps to Test

- Run unit-tests
- Run integration tests on all Gen3 devices

### References

[SC-100649](https://app.shortcut.com/particle/story/100649/gen2-supply-secure-device-os-4-x)

### Contributors
@technobly 
@keeramis 

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [x] Run unit/integration/application tests on device
- [x] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

**Review by XuGuohui — CHANGES_REQUESTED** [review]

File: hal/inc/spi_hal.h [hunk 1]
```diff
@@ -27,9 +27,6 @@
 typedef enum hal_spi_interface_t {
     HAL_SPI_INTERFACE1 = 0,    //maps to SPI1 (pins: A3, A4, A5)
     HAL_SPI_INTERFACE2 = 1     //maps to SPI3 (pins: D4, D3, D2)
-#if PLATFORM_ID == PLATFORM_ELECTRON_PRODUCTION
```
[the remark this answers is not in the document]
Comment (technobly) [reply hal/inc/spi_hal.h:30]: I changed this to:
```
/* Exported types ------------------------------------------------------------*/
typedef enum hal_spi_interface_t {
    HAL_SPI_INTERFACE1 = 0,    // maps to SPI (pins: A3, A4, A5)
    HAL_SPI_INTERFACE2 = 1,    // maps to SPI1 (pins: D4, D3, D2)
    HAL_SPI_INTERFACE3 = 2     // maps to SPI2 (pins: ?, ?, ?)
} hal_spi_interface_t;
```

File: hal/inc/usart_hal.h [hunk 2]
```diff
@@ -120,11 +120,6 @@ typedef struct hal_usart_ring_buffer_t {
 typedef enum hal_usart_interface_t {
     HAL_USART_SERIAL1 = 0,    //maps to USART_TX_RX
     HAL_USART_SERIAL2 = 1     //maps to USART_RGBG_RGBB
-#if PLATFORM_ID == PLATFORM_ELECTRON_PRODUCTION // Electron
```
[the remark this answers is not in the document]
Comment (technobly) [reply hal/inc/usart_hal.h:123]: I changed this to:
```
typedef enum hal_usart_interface_t {
    HAL_USART_SERIAL1 = 0,    // maps to USART_TX_RX
    HAL_USART_SERIAL2 = 1,    // maps to USART_RGBG_RGBB
    HAL_USART_SERIAL3 = 2,    // maps to ?
    HAL_USART_SERIAL4 = 3,    // maps to ?
    HAL_USART_SERIAL5 = 4     // maps to ?
} hal_usart_interface_t;
```

File: communication/src/build.mk [hunk 3]
```diff
@@ -42,14 +42,6 @@ CPPSRC += $(TARGET_SRC_PATH)/description.cpp
...
-CFLAGS += -DLOG_COMPILE_TIME_LEVEL=LOG_LEVEL_NONE
-endif
-
-ifeq ($(PLATFORM_ID),8)
-CFLAGS += -DLOG_COMPILE_TIME_LEVEL=LOG_LEVEL_NONE
-endif
-
 ifneq (,$(filter $(PLATFORM_ID), 13 23 25 26))
 ifneq ($(DEBUG_BUILD),y)
 ifneq ($(HYBRID_BUILD),y)
```
[the remark this answers is not in the document]
Comment (technobly) [reply communication/src/build.mk:47]: Since this appears to be more of a Gen3 thing, I'll open a separate story for it to evaluate.

**Review by XuGuohui — APPROVED** [review]

File: wiring/inc/spark_wiring_wifi.h [hunk 4]
```diff
@@ -251,45 +251,13 @@ class WiFiClass : public NetworkClass
...
-    int setHostname(const String& hostname)
-    {
-        return setHostname(hostname.c_str());
-    }
-
-    int setHostname(const char* hostname)
-    {
-        return network_set_hostname(*this, 0, hostname, nullptr);
-    }
-#endif // !HAL_PLATFORM_NCP
```
Comment (avtolstoy) [inline wiring/inc/spark_wiring_wifi.h:292]: Keep.

**Review by avtolstoy — CHANGES_REQUESTED** [review]

File: bootloader/src/main.c [hunk 5]
```diff
@@ -176,36 +168,13 @@ int main(void)
         features = 0xFF;        // ignore - corrupt. Top 4 bits should be the inverse of the bottom 4
     }
 
-#ifdef USE_LED_THEME
-    // Load LED theme colors
-    get_led_theme_colors(&FirmwareUpdateColor, &SafeModeColor, &DFUModeColor);
```
[the remark this answers is not in the document]
Comment (technobly) [reply bootloader/src/main.c:181]: I'll add HAL_PLATFORM_LED_THEME to hal_platform.h

File: communication/src/dsakeygen.cpp [hunk 6]
```diff
@@ -34,13 +34,8 @@
...
 #include "mbedtls_compat.h"
 #else
-# if PLATFORM_ID == 6 || PLATFORM_ID == 8
-#  include "wiced_security.h"
-#  include "crypto_open/bignum.h"
-# else
-#  include "tropicssl/rsa.h"
-# endif
-#endif
+#include "tropicssl/rsa.h"
```
[the remark this answers is not in the document]
Comment (technobly) [reply communication/src/dsakeygen.cpp:37]: Removed all `USE_MBEDTLS` conditionals and code that was defined under `!USE_MBEDTLS`.

File: crypto/inc/mbedtls_weaken.h [hunk 7]
```diff
@@ -91,74 +91,4 @@
 
 #pragma weak mbedtls_to_system_error
```
[the remark this answers is not in the document]
Comment (technobly) [reply crypto/inc/mbedtls_weaken.h:92]: Removed `mbedtls_weaken.h` and the code that was including it.  Removed CRYPTO_PART1_SIZE_OPTIMIZATIONS and files that used it like mbedtls_communication.inc and mbedtls_communication.cpp

File: newlib_nano/src/sources.mk [hunk 8]
```diff
@@ -1,8 +1,8 @@
 NEWLIBNANO_SRC_COMMON_PATH = $(NEWLIBNANO_MODULE_PATH)/src
 
-# Use FreeRTOS heap_4 and a set of newlib wrappers on Photon, P1 and Electron
-# Also on Argon, Boron, A SoM, B SoM, B5 SoM, Tracker
-ifneq (,$(filter $(PLATFORM_ID),6 8 10 12 13 22 23 25 26))
+# Use FreeRTOS heap_4 and a set of newlib wrappers on
+# Argon, Boron, A SoM, B SoM, B5 SoM, Tracker
+ifneq (,$(filter $(PLATFORM_ID),12 13 22 23 25 26))
```
[the remark this answers is not in the document]
Comment (technobly) [reply newlib_nano/src/sources.mk:5]: I like the first suggestion of using `HAL_PLATFORM_FREERTOS in malloc.cpp` and keeping it modular... I went with this as it's also the simpler change with least impact on code.

File: user/tests/wiring/no_fixture/uart.cpp [hunk 9]
```diff
@@ -17,75 +17,8 @@
 
 #include "application.h"
 #include "unit-test/unit-test.h"
-#include "Serial2/Serial2.h"
```
[the remark this answers is not in the document]
Comment (technobly) [reply user/tests/wiring/no_fixture/uart.cpp:20]: Moved the implementations to `spark_wiring_usarserial.h` and emptied the original headers.

File: build/version.mk [hunk 10]
```diff
@@ -1,7 +1,7 @@
-VERSION_STRING = 3.3.0
+VERSION_STRING = 4.0.0
```
[the remark this answers is not in the document]
Comment (technobly) [reply build/version.mk]: Reverted now... and created a proper bump in test/v4.0.0-alpha.1 that we can cherry pick back on top for testing after Keerthy pushes some changes for tests.

File: user/tests/wiring/api/wiring.cpp [hunk 11]
```diff
@@ -61,10 +61,6 @@ test(api_wiring_interrupt) {
 
     API_COMPILE(attachInterrupt(D0, &MyClass::handler, &myObj, RISING));
 
-#if PLATFORM_ID >= 6 && PLATFORM_ID <= 10
-    API_COMPILE(attachSystemInterrupt(SysInterrupt_TIM1_CC_IRQ, D0_callback));
```
[the remark this answers is not in the document]
Comment (keeramis) [reply user/tests/wiring/api/wiring.cpp]: @avtolstoy Should we remove `detachSystemInterrupt` as well as a part of this?

File: user/tests/wiring/no_fixture/interrupts.cpp [hunk 12]
```diff
@@ -82,142 +33,3 @@ class TestHandler
...
-	assertEqual(TestHandler::count, 1);
-	attachSystemInterrupt(SysInterrupt_SysTick, TestHandler()); // Override current handler
-	assertEqual(TestHandler::count, 1); // Previous handler has been destroyed
-	detachSystemInterrupt(SysInterrupt_SysTick);
-	assertEqual(TestHandler::count, 0);
-}
-
-test(INTERRUPTS_03_isisr_willpreempt_servicedirqn)
-{
-#if /* defined(STM32F10X_MD) || defined(STM32F10X_HD) || */ defined(STM32F2XX)
```
[the remark this answers is not in the document]
Comment (keeramis) [reply user/tests/wiring/no_fixture/interrupts.cpp]: @avtolstoy This is only defined for STM32. Should we make new tests for NRF platform? `SysInterrupt_SysTick` goes away with STM32

[the remark this answers is not in the document]
Comment (avtolstoy) [reply user/tests/wiring/no_fixture/interrupts.cpp]: We never adjusted it for Gen 3, it's just that. The main thing it's testing is some interrupts_hal functions. Just change it to use `attachInterruptDirect()`:
```c++
        static volatile bool cont = false;
	attachInterruptDirect(SysTick_IRQn, []() {
                detachInterruptDirect(SysTick_IRQn);
		assertTrue(HAL_IsISR());
		assertEqual((IRQn)HAL_ServicedIRQn(), SysTick_IRQn);
		cont = true;
	});
	while (!cont);
	assertFalse(HAL_WillPreempt(SysTick_IRQn, SysTick_IRQn));
	assertTrue(HAL_WillPreempt(NonMaskableInt_IRQn, SysTick_IRQn));
	assertFalse(HAL_WillPreempt(SysTick_IRQn, NonMaskableInt_IRQn));
```

File: user/tests/wiring/api/wiring.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (avtolstoy) [reply user/tests/wiring/api/wiring.cpp]: Yes. And the underlying implementation stubs.

**Review by avtolstoy — APPROVED** [review]

**Pull request merged**